### DISCUSSION TITLE
V5 Major Release

### DISCUSSION THREAD

**Discussion by mckalexee** [faceteer/facet#24] [Ideas]
The purpose of this discussion is to determine which features we want to add to the next major release.

## Potential Candidates

* Cross entity queries https://github.com/faceteer/facet/issues/23
* Data migration script https://github.com/faceteer/facet/issues/19
* Partial updates https://github.com/faceteer/facet/issues/18
* Separate out single/multiple query related functions https://github.com/faceteer/facet/pull/22
* Create an abstracted DynamoDB client interface https://github.com/faceteer/facet/issues/15

This is also an opportunity to change the interface in other ways if we've had different pain points.

**Comment by s-fletcher** [comment]
This would be a great time to add support for [DyanmoDB Transactions](https://github.com/faceteer/facet/issues/25).

**Reply by mckalexee** [reply]
That's a good idea.

**Reply by nturnage** [reply]
We're definitely going to want support for transactions from Tailwind's perspective. I think it makes faceteer that much stronger and versatile too.

**Reply by bill-kerr** [reply]
How do we imagine the API for this looking like? Off the top of my head, something like this might be ok:
```ts
const transaction = new Transaction([
  PostFacet.put({ id: '1' }),
  AccountFacet.put({ postCount: 1 }),
]);

await transaction.execute();
```

Obviously, the issue with this is that `Facet.put` or `Facet.delete` get executed immediately, so we'd need to change that API as well--to something like `Facet.put(...).execute()`.

...or we could make another method on the Facet that returns a "Transact Item" like this:

```ts
const transaction = new Transaction([
  PostFacet.transact.put({ id: '1' }),
  AccountFacet.transact.put({ postCount: 1 }),
]);

await transaction.execute();
```

I'm sure there might be other APIs we could explore as well.

**Reply by loganhuskins** [reply]
Additionally, [BatchWriteItem](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_BatchWriteItem.html) and [BatchGetItem](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_BatchGetItem.html) should probably be bundled in the same release and follow a similar API.

**Comment by bill-kerr** [comment]
Support for `Set` as an attribute would be a nice add, too. Should be pretty easy I think?

**Comment by SeriousBug** [comment]
I'd also love support for sparse GSIs. I think I know what changes are needed, I'll take a shot at it soon.
